### ISSUE TITLE
IntelliJ plugin removes capabilities from Xcode project

### ISSUE THREAD

**Issue by iitsoftware** [robovm/robovm#1092]
A newly created App ID in the Apple Dev Network contains by default Game Center and In-App-Purchase. This can NOT be disabled (also not edited, don't know why). In my Xcode project I have added these both capabilities. Whenever I restart IntelliJ or create an IPA, these capabilities were disabled by the plugin in the Xcode project and as a result I cannot deploy my app to the iPhone device with this error:

"The entitlements specified in your application’s Code Signing Entitlements file do not match those specified in your provisioning profile. (0xE8008016)."

**Comment by iitsoftware** [comment]
I'm using 1.5.0.

**Comment by badlogic** [comment]
The Xcode project RoboVM creates is not meant to be modified in any way. It
merely contains stubs for view controllers and references to storyboard
files and assets. If you need an Xcode project for actual code, you'll need
to create a separate one.

**Comment by iitsoftware** [comment]
But if I need to deploy the IPA for testing, this goes through the Xcode project (under windows/devices), right? That fails with the above error. In fact, it doesn’t seem that the IPA contains the same entitlements as the provisioning which has Game Center and In-App-Purchase enabled.

> Am 29.07.2015 um 16:55 schrieb Mario Zechner <notifications@github.com>:
> 
> The Xcode project RoboVM creates is not meant to be modified in any way. It
> merely contains stubs for view controllers and references to storyboard
> files and assets. If you need an Xcode project for actual code, you'll need
> to create a separate one.
> On Jul 29, 2015 4:25 PM, "iitsoftware" <notifications@github.com> wrote:
> 
> > I'm using 1.5.0.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/robovm/robovm/issues/1092#issuecomment-125969089>.
> >
> —
> Reply to this email directly or view it on GitHub.
> 






IIT Software GmbH
Fahrenheitstr. 13, D28359 Bremen, Germany
Tel: +49 421 330 46 088, Fax: +49 421 330 46 090
Amtsgericht Bremen, HRB 18624, Geschaeftsfuehrer: Andreas Mueller
Steuernummer: 460/118/06404 VAT: DE199945912

**Comment by badlogic** [comment]
No, there's no Xcode project involved when building an IPA. You need to add
the path to your entitlements.plist (stored outside the Xcode project) to
your robovm.xml. See
http://docs.robovm.com/configuration.html#-lt-iosentitlementsplist-gt

**Comment by iitsoftware** [comment]
Sorry, but you didn’t understand what I meant.

I did create an IPA with RoboVM. RoboVM asked me about the provisioning which I have downloaded. Ok so far. The IPA has been created.

But now I want to deploy that on my device which is connected to my Mac. There is a function in Xcode to do that under Windows/Device. If I now select the IPA and Xcode tries to deploy, I’m getting the entitlement error as stated below.

If that is not the proper way to deploy the IPA to a connected device, please let me know a different way. 

> Am 29.07.2015 um 17:36 schrieb Mario Zechner <notifications@github.com>:
> 
> No, there's no Xcode project involved when building an IPA. You need to add
> the path to your entitlements.plist (stored outside the Xcode project) to
> your robovm.xml. See
> http://docs.robovm.com/configuration.html#-lt-iosentitlementsplist-gt
> On Jul 29, 2015 4:59 PM, "iitsoftware" <notifications@github.com> wrote:
> 
> > But if I need to deploy the IPA for testing, this goes through the Xcode
> > project (under windows/devices), right? That fails with the above error. In
> > fact, it doesn’t seem that the IPA contains the same entitlements as the
> > provisioning which has Game Center and In-App-Purchase enabled.
> >
> > > Am 29.07.2015 um 16:55 schrieb Mario Zechner <notifications@github.com>:
> > >
> > > The Xcode project RoboVM creates is not meant to be modified in any way.
> > It
> > > merely contains stubs for view controllers and references to storyboard
> > > files and assets. If you need an Xcode project for actual code, you'll
> > need
> > > to create a separate one.
> > > On Jul 29, 2015 4:25 PM, "iitsoftware" <notifications@github.com> wrote:
> > >
> > > > I'm using 1.5.0.
> > > >
> > > > —
> > > > Reply to this email directly or view it on GitHub
> > > > <https://github.com/robovm/robovm/issues/1092#issuecomment-125969089>.
> > > >
> > > —
> > > Reply to this email directly or view it on GitHub.
> > >
> >
> >
> >
> >
> >
> >
> > IIT Software GmbH
> > Fahrenheitstr. 13, D28359 Bremen, Germany
> > Tel: +49 421 330 46 088, Fax: +49 421 330 46 090
> > Amtsgericht Bremen, HRB 18624, Geschaeftsfuehrer: Andreas Mueller
> > Steuernummer: 460/118/06404 VAT: DE199945912
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/robovm/robovm/issues/1092#issuecomment-125980569>.
> >
> —
> Reply to this email directly or view it on GitHub.
> 

-- 
Andreas Mueller
IIT Software GmbH, Bremen/Germany
http://www.swiftmq.com





IIT Software GmbH
Fahrenheitstr. 13, D28359 Bremen, Germany
Tel: +49 421 330 46 088, Fax: +49 421 330 46 090
Amtsgericht Bremen, HRB 18624, Geschaeftsfuehrer: Andreas Mueller
Steuernummer: 460/118/06404 VAT: DE199945912

**Comment by iitsoftware** [comment]
Here is a screenshot of the error:



IIT Software GmbH
Fahrenheitstr. 13, D28359 Bremen, Germany
Tel: +49 421 330 46 088, Fax: +49 421 330 46 090
Amtsgericht Bremen, HRB 18624, Geschaeftsfuehrer: Andreas Mueller
Steuernummer: 460/118/06404 VAT: DE199945912

**Comment by BlueRiverInteractive** [comment]
Have you tried installing the IPA via iTunes? This is the way I test my IPA's on test devices.

A) First of all make sure that the provisioning profile is AdHoc and it contains the UDID of your test device(s).
B) Open up iTunes and switch to the apps section.
C) Drag and drop your IPA onto that section.
D) Connect your device.
E) Switch to your device in iTunes and select apps.
F) Press the Install button next to your app and then Synchronize.
G) If the provisioning profile was correct, your app will now be installed onto your device. If it stops at Installing... then the profile doesn't contain the UDID of the test device.

**Comment by iitsoftware** [comment]
Yes, tried that and it stop st "Installing ...". The provision contains the UUID, at least that was what I have specified. 



> Am 29.07.2015 um 19:03 schrieb BlueRiverInteractive <notifications@github.com>:
> 
> Have you tried installing the IPA via iTunes? This is the way I test my IPA's on test devices.
> 
> A) First of all make sure that the provisioning profile is AdHoc and it contains the UDID of your test device(s).
> B) Open up iTunes and switch to the apps section.
> C) Drag and drop your IPA onto that section.
> D) Connect your device.
> E) Switch to your device in iTunes and select apps.
> F) Press the Install button next to your app and then Synchronize.
> G) If the provisioning profile was correct, your app will now be installed onto your device. If it stops at Installing... then the profile doesn't contain the UDID of the test device.
> 
> —
> Reply to this email directly or view it on GitHub.
> 


IIT Software GmbH
Fahrenheitstr. 13, D28359 Bremen, Germany
Tel: +49 421 330 46 088, Fax: +49 421 330 46 090
Amtsgericht Bremen, HRB 18624, Geschaeftsfuehrer: Andreas Mueller
Steuernummer: 460/118/06404 VAT: DE199945912

**Comment by BlueRiverInteractive** [comment]
Maybe the new provisioning profile hasn't been applied yet? 
When you have updated them you also have to synchronize your Xcode to make them work. 
In Xcode go to Preferences, select your Apple ID and then "View Details...". You should see all your provisioning profiles and "Expires At" should match with the entries in the apple developer dashboard. To refresh the profiles simply press the refresh/arrow button.

**Comment by iitsoftware** [comment]
1) The profile has been applied (I do a refresh all times). It is iOS development (= ad hoc?).
2) RoboVM asks about the profile when generating the IPA. It is the correct one.
3) Checking the profile I see my correct UDID:

	<key>ProvisionedDevices</key>
	<array>
		<string>7283877a11975b6cbca80496e6d170a2e699f835</string>
	</array>

4) If I try to install it with iTunes, it copies it to the device but hangs on „Installing …“
5) If I try to install it with Xcode (Windows->Devices), it complains about not matching Entitlements.

The latter seems to be a bug in RoboVM, that’s the reason why I filed this here. The provisioning profiles per default enables Game Center and In-App-Purchase. This can for whatever reason not be disabled in the profile at the Apple Dev Network. So this is enabled. RoboVM does not enable it and I don’t know how to force it to do so. There is no Entitlement.plist I can specify. Xcode 6.4 does not generate this file and I don’t find any hint in the Internet how to create it.

So this is quite frustating as this is an important step to deploy it on the device and test it. It would be great if RoboVM would had some more detailed docs about it. But all I find is a link to the Apple Dev Connection…

What I need is a) How to create the Entitlement.plist or b) Why does it stop on „Installing…“ or c) A workaround to finally deploy my app. Otherwise I can simply stop development.

-- 
Andreas Mueller
IIT Software GmbH, Bremen/Germany
http://www.swiftmq.com

> Am 29.07.2015 um 20:09 schrieb BlueRiverInteractive <notifications@github.com>:
> 
> Maybe the new provisioning profile hasn't been applied yet? 
> When you have updated them you also have to synchronize your Xcode to make them work. 
> In Xcode go to Preferences, select your Apple ID and then "View Details...". You should see all your provisioning profiles and "Expires At" should match with the entries in the apple developer dashboard. To refresh the profiles simply press the refresh/arrow button.
> 
> —
> Reply to this email directly or view it on GitHub.
> 






IIT Software GmbH
Fahrenheitstr. 13, D28359 Bremen, Germany
Tel: +49 421 330 46 088, Fax: +49 421 330 46 090
Amtsgericht Bremen, HRB 18624, Geschaeftsfuehrer: Andreas Mueller
Steuernummer: 460/118/06404 VAT: DE199945912

**Comment by badlogic** [comment]
OK, here's how i created an app id, ad-hoc provisioning profile, and deployed the app to my ipad via itunes.

## Creating the App
I used IntelliJ IDEA to create the app, based on the simple single view template:

![http://sht.tl/wZKa6F](http://sht.tl/wZKa6F)

Note the application id `com.badlogicgames.myiosapp`. This is also the bundle id of my app, which can be found in `robovm.properties` once the project is generated:

```
app.version=1.0
app.id=com.badlogicgames.myiosapp
app.mainclass=com.badlogicgames.myiosapp.Main
app.executable=Main
app.build=1
app.name=My iOS App
```

That will be substituted by RoboVM in the `Info.plist` file when the app is build (either via a run configuration for device/simulator or when building an IPA). You can see the substitution in `Info.plist`:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
    <dict>
        <key>CFBundleDevelopmentRegion</key>
        <string>en</string>
        <key>CFBundleDisplayName</key>
        <string>${app.name}</string>
        <key>CFBundleExecutable</key>
        <string>${app.executable}</string>
        <key>CFBundleIdentifier</key>
        <string>${app.id}</string>
...
```

## Creating the App ID
Next i headed to the Apple Developer Center and clicked on **Identifiers -> App IDs**, that will list my existing app ids and allows me to create a new one:

![http://sht.tl/74W6D7](http://sht.tl/74W6D7)

I click on the plus button in the top right corner to create a new App ID for my new project, and fill out the details:

![http://sht.tl/9fJVIL](http://sht.tl/9fJVIL)

The App ID description is just that, a description for the ID, shown when its listed in the developer center. Next i entered the Bundle ID, since this is an explicit app id meant for my app. I leave everything else as is. As you said, Game Center and In-App Purchase are enabled by default in App Services.

I click **Continue** then **Submit** and now have a new App ID:

![http://sht.tl/P4D7pl](http://sht.tl/P4D7pl)

## Creating the Provisioning Profile
Next i create a provisioning profile for my new App ID. I click on **Provisioning Profiles -> All**, then click the plus button in the top right corner:

![http://sht.tl/OFXlH8](http://sht.tl/OFXlH8)

I create an Ad Hoc provisioning profile:

![http://sht.tl/6lQmvo](http://sht.tl/6lQmvo)

Then select my app id:

![http://sht.tl/6FbSBr](http://sht.tl/6FbSBr)

I select the certificate with which my app will get signed:

![http://sht.tl/SYgsYd](http://sht.tl/SYgsYd)

Then i select the devices to which i'll be allowed to install the Ad Hoc IPA:

![http://sht.tl/XW7KSB](http://sht.tl/XW7KSB)

Finally i specify the profile name:

![http://sht.tl/IRGhDf](http://sht.tl/IRGhDf)

I can now download the provisioning profile and install it by double clicking the downloaded file:

![http://sht.tl/TAuxHs](http://sht.tl/TAuxHs)

Alternatively i could open Xcode, go to **Xcode -> Preferences -> Accounts**, click **Details** and then click the refresh button to download the latest provisioning profiles.

![http://sht.tl/42YEWt](http://sht.tl/42YEWt)

Which looks like this after the refresh:

![http://sht.tl/UMVUC1](http://sht.tl/UMVUC1)

Note that all the distribution profiles have the In-App Purchase and GameCenter entitlements.

## Creating the IPA
I switch back to IntelliJ IDEA with my project open. I select **Build -> Create IPA**, in Eclipse i'd right click the project, and select **RoboVM -> Create IPA**. Next i'm shown a dialogue that let's me select my signing identify and provisioning profile among other things:

![http://sht.tl/FDYHmp](http://sht.tl/FDYHmp)

I select a destionation folder into which my IPA will be written. Next i select the signing identity that's I associated with my provisioning profile earlier (certificate). Finally i select the provisioning profile for my app and click **OK**.

You can now watch the RoboVM compiler do it's work in the RoboVM Console (available both in IntelliJ IDEA and Eclipse. The interesting bit is the signing step, where the entitlements are applied:

```
[DEBUG] Copying AdHoc provisioning profile: My iOS App Provisioning Profile (7F5Y92G2Z4.com.badlogicgames.myiosapp)
[DEBUG] Code signing app using identity 'iPhone Distribution: Mario Zechner (7F5Y92G2Z4)' with fingerprint XXXREDACTEDFINGERPRINTXXX
[DEBUG] codesign -f -s XXXREDACTEDFINGERPRINTXXX --entitlements /Users/badlogic/Library/Caches/IdeaIC15/plugins-sandbox/system/tmp/robovm8808748861012410547.tmp/Entitlements.plist /Users/badlogic/Desktop/test/Main.app
```

As you can see, RoboVM generates an Entitlements.plist file and puts it in a temporary folder. The contents of that file are as follows:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>get-task-allow</key>
	<false/>
	<key>keychain-access-groups</key>
	<array>
		<string>7F5Y92G2Z4.*</string>
	</array>
	<key>application-identifier</key>
	<string>7F5Y92G2Z4.com.badlogicgames.myiosapp</string>
	<key>com.apple.developer.team-identifier</key>
	<string>7F5Y92G2Z4</string>
</dict>
</plist>
```

## Installing the IPA via iTunes
Next i connect my device (which i enabled for ad hoc distribution when creating the provisioning profile) and start iTunes. I select my device and view the **Apps**:

![http://sht.tl/cGYblY](http://sht.tl/cGYblY)

I open up the folder to which i compiled the IPA and double click the IPA:

![http://sht.tl/S5nFhV](http://sht.tl/S5nFhV)

This adds it to my device's apps in iTunes, ready to be installed to the connected device:

![http://sht.tl/Ee7kqs](http://sht.tl/Ee7kqs)

I click on **Install** next to the app name, then **Apply** at the bottom right corner:

![http://sht.tl/HzJskc](http://sht.tl/HzJskc)

This now installs the IPA to my device:

![http://sht.tl/ktn1xx](http://sht.tl/ktn1xx)

## What about the Entitlements.plist file?
For this example *we don't need a custom plist file! The entitlements for In-App Purchases and GameCenter are automatically added by Apple, not by RoboVM nor manually*. If you get an error, it's likely something in your setup deviates from the above.

If you need additional entitlements, e.g. iAds or similar services, you can take the above `Entitlements.plist` example and simply add the additional entitlements. Here's an example with healthkit, IAP, maps etc.:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>get-task-allow</key>
	<false/>
	<key>keychain-access-groups</key>
	<array>
		<string>7F5Y92G2Z4.*</string>
	</array>
	<key>application-identifier</key>
	<string>7F5Y92G2Z4.com.badlogicgames.myiosapp</string>
	<key>com.apple.developer.team-identifier</key>
	<string>7F5Y92G2Z4</string>
        <key>com.apple.developer.healthkit</key>
	<true/>
	<key>com.apple.developer.homekit</key>
	<true/>
	<key>com.apple.developer.icloud-container-identifiers</key>
	<array/>
	<key>com.apple.developer.in-app-payments</key>
	<array/>
</dict>
</plist>
```

You can find these entitlement keys and their potential values in the [Apple Developer Documentation](https://developer.apple.com/library/ios/documentation/Miscellaneous/Reference/EntitlementKeyReference/Chapters/AboutEntitlements.html).

I'd suggest you give this little tutorial a quick try to validate that it's working for you. Especially the signing identity and provisioning profile setup. Once that works for you, try to apply it to your own application.

**Comment by iitsoftware** [comment]
Thanks for the quick tutorial. But it still hangs in „Install…“.

I’ve cross checked everything. Here are some screenshots:

App ID:




Certs:



Device:



Profiles:



Ad Hoc Profile:



Xcode after refresh:



robovm.properties:

app.version=1.0
app.id=com.swiftmq.admin.ios.swiftmq
app.mainclass=com.swiftmq.admin.ios.explorer.Main
app.executable=Main
app.build=1
app.name=SwiftMQ Explorer

robovm.xml:

<config>
  <executableName>${app.executable}</executableName>
  <mainClass>${app.mainclass}</mainClass>
  <os>ios</os>
  <arch>thumbv7</arch>
  <resources>
    <resource>
      <directory>resources</directory>
    </resource>
  </resources>
  <target>ios</target>
  <iosInfoPList>Info.plist.xml</iosInfoPList>
  <forceLinkClasses>
    <pattern>com.android.okhttp.HttpHandler</pattern>
    <pattern>com.android.org.conscrypt.OpenSSLProvider</pattern>
    <pattern>org.apache.harmony.security.provider.cert.DRLCertFactory</pattern>
    <pattern>com.android.org.bouncycastle.jce.provider.BouncyCastleProvider</pattern>
    <pattern>org.apache.harmony.security.provider.crypto.CryptoProvider</pattern>
    <pattern>com.android.org.conscrypt.JSSEProvider</pattern>
    <pattern>com.android.org.conscrypt.OpenSSLRandom</pattern>
    <pattern>com.android.org.conscrypt.OpenSSLMessageDigestJDK$MD5</pattern>
    <pattern>com.swiftmq.net.PlainSocketFactory</pattern>
    <pattern>com.swiftmq.net.JSSESocketFactory</pattern>
    <pattern>com.swiftmq.auth.ChallengeResponseFactoryImpl</pattern>
  </forceLinkClasses>
</config>

Create IPS dialog:



Create IPA output:

[DEBUG] Installing Info.plist to /Users/am/Development/explorer-ios/Main.app
[DEBUG] xcrun dsymutil -o /Users/am/Development/explorer-ios/Main.app.dSYM /Users/am/Development/explorer-ios/Main.app/Main
[DEBUG] xcrun strip -x /Users/am/Development/explorer-ios/Main.app/Main
[DEBUG] Copying AdHoc provisioning profile: Ad Hoc iPhone 6 (34W5YCX76B.com.swiftmq.admin.ios.swiftmq)
[DEBUG] Code signing using identity 'iPhone Distribution: IIT Software GmbH (34W5YCX76B)' with fingerprint 904B1E14DB855189FA3C0512DDCD59102D33FD84
[DEBUG] codesign -f -s 904B1E14DB855189FA3C0512DDCD59102D33FD84 --entitlements /Users/am/Library/Caches/IntelliJIdea14/tmp/robovm1984570103651907421.tmp/Entitlements.plist /Users/am/Development/explorer-ios/Main.app
[DEBUG] ln -f -s _CodeSignature/CodeResources /Users/am/Development/explorer-ios/Main.app/CodeResources
[DEBUG] /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/usr/bin/PackageApplication /Users/am/Development/explorer-ios/Main.app -o /Users/am/Development/explorer-ios/Main.ipa
[INFO] Package successfully created in /Users/am/Development/explorer-ios
[INFO] IPA creation complete, 0 errors, 0 warnings

iTunes install:



The app is copied to the iPhone. When I press the icon I see this and it never changes:







IIT Software GmbH
Fahrenheitstr. 13, D28359 Bremen, Germany
Tel: +49 421 330 46 088, Fax: +49 421 330 46 090
Amtsgericht Bremen, HRB 18624, Geschaeftsfuehrer: Andreas Mueller
Steuernummer: 460/118/06404 VAT: DE199945912

**Comment by badlogic** [comment]
I'm afraid the images didn't come through. Please send a mail to mario at robovm dot com for a follow up.

**Comment by badlogic** [comment]
After a quick one-on-one it turned out that it's a problem with the project setup. Andreas, if you don't mind, i'll close this one. Feel free to post here if you continue to have problems.

**Comment by iitsoftware** [comment]
After creating a new project and after copying all necessary files to it, it works. Seems my project was damaged in some way!

> Am 30.07.2015 um 14:36 schrieb Mario Zechner <notifications@github.com>:
> 
> Closed #1092.
> 
> —
> Reply to this email directly or view it on GitHub.
> 

-- 
Andreas Mueller
IIT Software GmbH, Bremen/Germany
http://www.swiftmq.com





IIT Software GmbH
Fahrenheitstr. 13, D28359 Bremen, Germany
Tel: +49 421 330 46 088, Fax: +49 421 330 46 090
Amtsgericht Bremen, HRB 18624, Geschaeftsfuehrer: Andreas Mueller
Steuernummer: 460/118/06404 VAT: DE199945912

**Comment by badlogic** [comment]
Awesome, glad we could resolve it!
